File: problems/chapter11/miiitomi/003.cpp
// AtCoder ABC049   D - 連結
// https://atcoder.jp/contests/abc049/tasks/arc065_b

#include <iostream>
#include <vector>
#include <map>
using namespace std;

struct UnionFind {
    vector<int> par, siz;

    UnionFind(int n) : par(n, -1), siz(n, 1) {}

    int root(int x) {
        if (par[x] == -1) return x;
        else return par[x] = root(par[x]);
    }

    bool issame(int x, int y) {
        return root(x) == root(y);
    }

    bool unite(int x, int y) {
        x = root(x);
        y = root(y);
        if (x == y) return false;
        if (siz[x] < siz[y]) swap(x, y);
        par[y] = x;
        siz[x] += siz[y];
        return true;
    }

    int size(int x) {
        return siz[root(x)];
    }
};

int main() {
    int N, K, L;
    cin >> N >> K >> L;

    UnionFind uf1(N);
    for (int i = 0; i < K; i++) {
        int p, q;
        cin >> p >> q;
        p--;
        q--;
        uf1.unite(p, q);
    }

    UnionFind uf2(N);
    for (int j = 0; j < L; j++) {
        int r, s;
        cin >> r >> s;
        r--;
        s--;
        uf2.unite(r, s);
    }

    vector<pair<int, int>> V(N);
    for (int i = 0; i < N; i++) {
        pair<int, int> p = make_pair(uf1.root(i), uf2.root(i));
        V[i] = p;
    }
    map<pair<int, int>, int> M;
    for (int i = 0; i < N; i++) {
        if (M.count(V[i])) {
            M[V[i]] += 1;
        } else {
            M[V[i]] = 1;
        }
    }

    for (int i = 0; i < N; i++) {
        if (i < N-1) {
            cout << M[V[i]] << " ";
        } else {
            cout << M[V[i]] << endl;
        }
    }
}

File: problems/chapter11/mshibatatt/003.py
# https://atcoder.jp/contests/abc049/tasks/arc065_b

from typing import Counter


class UnionFind():
    def __init__(self, n) -> None:
        self.par = [-1] * n
        self.siz = [1] * n
    
    def root(self, x):
        if self.par[x] == -1:
            return x
        else:
            self.par[x] = self.root(self.par[x])
            return self.par[x]
    
    def issame(self, x, y):
        return self.root(x) == self.root(y)

    def unite(self, x, y):
        x = self.root(x)
        y = self.root(y)

        if x == y:
            return False
        if self.siz[x] < self.siz[y]:
            x, y = y, x
        
        self.par[y] = x
        self.siz[x] += self.siz[y]
        return True

    def size(self, x):
        return self.siz[self.root(x)]

def main():
    N, K, L = map(int, input().split())
    roads = []
    rails = []
    for _ in range(K):
        a, b = map(int, input().split())
        roads.append([a-1, b-1])

    for _ in range(L):
        a, b = map(int, input().split())
        rails.append([a-1, b-1])

    output = []
    uf_road = UnionFind(N)
    uf_rail = UnionFind(N)

    for i in roads:
        uf_road.unite(i[0], i[1])
    for i in rails:
        uf_rail.unite(i[0], i[1])
    
    counter_dict = {}
    roots = []
    for i in range(N):
        roots.append((uf_road.root(i), uf_rail.root(i)))
        if roots[i] in counter_dict.keys():
            counter_dict[roots[i]] += 1
        else:
            counter_dict[roots[i]] = 1
    
    output = []
    for i in roots:
        output.append(counter_dict[i])

    print(' '.join(map(str, output)))

if __name__=='__main__':
    main()